clients/src/test/java/org/apache/kafka/clients/consumer/ConsumerRecordsTest.java

```diff
@@ -31,32 +31,129 @@
-        records.put(new TopicPartition(topic, 2), new ArrayList<>());
-
-        ConsumerRecords<Integer, String> consumerRecords = new ConsumerRecords<>(records);
-        Iterator<ConsumerRecord<Integer, String>> iter = consumerRecords.iterator();
-
-        int c = 0;
-        for (; iter.hasNext(); c++) {
-            ConsumerRecord<Integer, String> record = iter.next();
-            assertEquals(1, record.partition());
             assertEquals(topic, record.topic());
-            assertEquals(c, record.offset());
+            assertEquals(currentPartition, record.partition());
+            assertEquals(recordCount % recordSize, record.offset());
+            assertEquals(recordCount % recordSize, record.key());
+            assertEquals(String.valueOf(recordCount % recordSize), record.value());
+
+            recordCount++;
         }
-        assertEquals(2, c);
+
+        // Including empty partition
+        assertEquals(partitionSize, partitionCount + 1);
+    }
+
+    @Test
+    public void testRecordsWithNullTopic() {
+        String nullTopic = null;
+        ConsumerRecords<Integer, String> consumerRecords = ConsumerRecords.empty();
+        IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> consumerRecords.records(nullTopic));
+        assertEquals("Topic must be non-null.", exception.getMessage());
+    }
+
+
+    @Test
+    public void testRecords() {
+        List<String> topics = Arrays.asList("topic1", "topic2", "topic3", "topic4");
+        int recordSize = 3;
+        int partitionSize = 10;
+        int emptyPartitionIndex = 6;
+        int expectedTotalRecordSizeOfEachTopic = recordSize * (partitionSize - 1);
+
+        ConsumerRecords<Integer, String> consumerRecords = buildTopicTestRecords(recordSize, partitionSize, emptyPartitionIndex, topics);
+
+        for (String topic : topics) {
+            Iterable<ConsumerRecord<Integer, String>> records = consumerRecords.records(topic);
+            Iterator<ConsumerRecord<Integer, String>> iterator = records.iterator();
+            int recordCount = 0;
+            int partitionCount = 0;
+            int currentPartition = -1;
+
+            while (iterator.hasNext()) {
+                ConsumerRecord<Integer, String> record = iterator.next();
+
+                if (record.partition() == emptyPartitionIndex) {
+                    fail("Partition " + emptyPartitionIndex + " is not empty");
+                }
+
+                // Check if we have moved to a new partition
+                if (currentPartition != record.partition()) {
+                    // Increment the partition count as we have encountered a new partition
+                    partitionCount++;
+                    // Update the current partition to the new partition
+                    currentPartition = record.partition();
+                }
+
+                assertEquals(topic, record.topic());
```

Comment:
those checks are existent in both test cases. Maybe we can reuse them?

Reply:
Cool, it will be more clear